### PR TITLE
feat(dropbox): add users sync

### PR DESCRIPTION
**Pull request by athielba**
## Description

This branch will be the parent for all the upcoming branches, In this branch

1. Authentication & Authorization flow
2. Refresh auth token for all organizations (every 3.5 hours)
3. Fetch users
4. sync uses with Elba

## Checklist:

Before you create this PR, confirm all the requirements listed below by checking the checkboxes like this (`[x]`).

- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests to cover the new feature or fixes.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,3 +1,16 @@
...
+      "source.fixAll": "never"
+    }
+  },
+  "editor.codeActionsOnSave": {
+    "source.fixAll": "explicit"
+  },
+  "typescript.tsdk": "node_modules/typescript/lib",
+  "editor.formatOnSave": true,
+  "files.trimTrailingWhitespace": true
+}
```
Comment (elbalexandre) [inline .vscode/settings.json:16]: Some of theses changes could be handy, but some are related to your own configuration. It could be discuted in a dedicated PR.

File: .vscode/workspace-settings.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  },
+  "editor.codeActionsOnSave": {
+    "source.fixAll.eslint": true
+  },
+  "editor.formatOnSave": true,
+  "editor.tabSize": 2,
+  "eslint.validate": ["typescript", "typescriptreact"],
+  "files.trimTrailingWhitespace": true,
+  "prettier.requireConfig": true
+}
```
Comment (elbalexandre) [inline .vscode/workspace-settings.json:31]: same here

File: apps/dropbox/drizzle/0000_quiet_sauron.sql [hunk 3]
```diff
@@ -0,0 +1,23 @@
+CREATE TABLE IF NOT EXISTS "shared_links" (
+	"url" text NOT NULL,
+	"organisation_id" uuid NOT NULL,
+	"team_member_id" text NOT NULL,
+	"link_access_level" text NOT NULL,
+	"path_lower" text NOT NULL,
+	CONSTRAINT shared_links_url_path_lower_pk PRIMARY KEY("url","path_lower")
+);
```
Comment (elbalexandre) [inline apps/dropbox/drizzle/0000_quiet_sauron.sql:8]: We might not need shared_links for users sync

File: apps/dropbox/drizzle/0000_quiet_sauron.sql [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+--> statement-breakpoint
+CREATE TABLE IF NOT EXISTS "tokens" (
```
Comment (elbalexandre) [inline apps/dropbox/drizzle/0000_quiet_sauron.sql:10]: It might be less missleading if every integration store organisation data in a `organisations` table.
With this name `organisation_id` could be `id`, it make more sense for a pk.

File: apps/dropbox/drizzle/0000_quiet_sauron.sql [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	"organisation_id" uuid PRIMARY KEY NOT NULL,
+	"access_token" text NOT NULL,
+	"refresh_token" text NOT NULL,
+	"team_name" text NOT NULL,
+	"admin_team_member_id" text NOT NULL,
+	"root_namespace_id" text NOT NULL,
+	"unauthorized_at" timestamp,
+	"refresh_after" timestamp,
+	"expires_at" timestamp,
```
Comment (elbalexandre) [inline apps/dropbox/drizzle/0000_quiet_sauron.sql:19]: What are those for ?

File: apps/dropbox/next.config.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+/** @type {import('next').NextConfig} */
+const nextConfig = {
+  ignoreDuringBuilds: true,
+  webpack: (config, { webpack }) => {
+    config.plugins.push(
+      new webpack.IgnorePlugin({
+        resourceRegExp: /^cloudflare:sockets$/,
```
Comment (elbalexandre) [inline apps/dropbox/next.config.js:7]: Why do we need this ?

File: apps/dropbox/package.json [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    "test": "vitest",
+    "big-bang": "pnpm database:down && pnpm test:database:down && pnpm database:up && pnpm test:database:up && pnpm migrate",
+    "test:database:up": "docker-compose -p dropbox-test --env-file=.env.test.local up -d",
+    "test:database:down": "docker-compose -p dropbox-test down --volumes",
+    "database:down": "docker-compose -p dropbox down --volumes",
+    "database:up": "docker-compose -p dropbox --env-file=.env.local up -d",
+    "migration:generate": "drizzle-kit generate:pg --schema=./src/database/schema.ts",
+    "migration:push": "npx tsx ./src/database/migrate.ts .env.local && npx tsx ./src/database/migrate.ts .env.test.local",
+    "migrate": "pnpm migration:generate && pnpm migration:push",
+    "inngest": "npx inngest-cli@latest dev -u http://localhost:3300/api/inngest"
```
Comment (elbalexandre) [inline apps/dropbox/package.json:19]: Could be great to have the same scripts as the template, and add custom script at the end.

File: apps/dropbox/package.json [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  },
+  "dependencies": {
+    "@elba-security/schemas": "workspace:*",
+    "@elba-security/sdk": "workspace:*",
+    "@t3-oss/env-nextjs": "^0.7.1",
```
Comment (elbalexandre) [inline apps/dropbox/package.json:24]: Do we need to make a front-end here ? otherwise zod is just a faster way to validate env variable.

File: apps/dropbox/src/app/api/inngest/functions/tokens/data.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+export const getOrganisationRefreshToken = async (organisationId: string) => {
+  try {
+    return await db
+      .select({
+        refreshToken: tokens.refreshToken,
+      })
+      .from(tokens)
+      .where(and(eq(tokens.organisationId, organisationId)));
+  } catch (error) {
+    throw Error(
```
Comment (elbalexandre) [inline apps/dropbox/src/app/api/inngest/functions/tokens/data.ts:19]: In a inngest context, this should be a nonretriable error.

**Review by elbalexandre — COMMENTED** [review]
We might prefer to follow the template code structure (connectors + inngest).

File: apps/dropbox/src/database/migrate.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+if (!envFileName) {
+  throw new Error(
+    'Could not find environment file. Please specify the path to the environment file as the first argument.\nUsage: npx ts-node ./src/database/migrate.ts <env-file>'
+  );
+}
+
+config({ path: envFileName });
+
+const client = postgres(env.POSTGRES_URL!);
+
```
Comment (elbalexandre) [inline apps/dropbox/src/database/migrate.ts:18]: IMO migration should not be part of the source code

File: apps/dropbox/src/database/client.ts [hunk 11]
```diff
@@ -0,0 +1,15 @@
+import { drizzle } from 'drizzle-orm/postgres-js';
+import postgres from 'postgres';
+import { env } from '@/common/env';
+
+const connectionString = () => {
```
Comment (elbalexandre) [inline apps/dropbox/src/database/client.ts:5]:
```suggestion
const getConnectionString = () => {
```

File: apps/dropbox/src/common/clients/inngest.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import { EventSchemas, Inngest } from 'inngest';
+import { zodEventSchemas } from './types';
+import { unauthorizedMiddleware } from '@/app/api/inngest/middlewares/unauthorized-middleware';
+import { rateLimitMiddleware } from '@/app/api/inngest/middlewares/rate-limit-middleware';
+
+export type FunctionHandler = Parameters<typeof inngest.createFunction>[2];
+
+export const inngest = new Inngest({
+  id: 'dropbox',
+  schemas: new EventSchemas().fromZod(zodEventSchemas),
```
Comment (elbalexandre) [inline apps/dropbox/src/common/clients/inngest.ts:10]: We prefer to use ts type instead of zod as it does not bring any value.

File: apps/dropbox/src/common/clients/inngest.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+  middleware: [unauthorizedMiddleware, rateLimitMiddleware],
```
Comment (elbalexandre) [inline apps/dropbox/src/common/clients/inngest.ts:11]: The logger should be configured as well.

File: apps/dropbox/src/common/clients/elba.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
+import { Elba } from '@elba-security/sdk';
+import { env } from '../env';
+
+export const elbaAccess = ({
```
Comment (elbalexandre) [inline apps/dropbox/src/common/clients/elba.ts:4]:
```suggestion
export const getElba = ({
```

File: apps/dropbox/src/app/oauth/service.ts [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    });
+
+    await inngest.send({
+      name: 'users/run-user-sync-jobs',
+      data: {
+        organisationId,
+        isFirstSync: true,
+        syncStartedAt: new Date().toISOString(),
+      },
+    });
```
Comment (elbalexandre) [inline apps/dropbox/src/app/oauth/service.ts:95]: This could be batched in a single inngest.send call

File: apps/dropbox/src/app/oauth/service.ts [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+    throw new Error('Could not get root namespace id');
+  }
+
+  try {
+    const tokenExpiresAt = addSeconds(new Date(), expires_in);
+    await insertAccessToken({
+      organisationId,
+      accessToken,
+      refreshToken: refresh_token,
+      expiresAt: tokenExpiresAt,
```
Comment (elbalexandre) [inline apps/dropbox/src/app/oauth/service.ts:72]: Why do we need to store this ?

File: apps/dropbox/src/app/api/inngest/functions/users/run-user-sync-jobs.ts [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+    const { members, ...rest } = await dbx.fetchUsers(cursor);
+
+    if (members.length > 0) {
+      await elba.users.update({
+        users: members,
+      });
+    }
+
+    return rest;
```
Comment (elbalexandre) [inline apps/dropbox/src/app/api/inngest/functions/users/run-user-sync-jobs.ts:39]: Perhaps this can returns only `rest.hasMore` as it's the only field used.

File: apps/dropbox/src/app/api/inngest/functions/users/run-user-sync-jobs.ts [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  await step.run('user-sync-finalize', async () => {
+    await elba.users.delete({
+      syncedBefore: syncStartedAt,
+    });
+  });
+
+  return {
+    success: true,
```
Comment (elbalexandre) [inline apps/dropbox/src/app/api/inngest/functions/users/run-user-sync-jobs.ts:60]: In the inngest console we already have a success status by default; Perhaps returning the status 'ongoing' | 'completed' will be more usefull.

File: apps/dropbox/src/app/api/inngest/functions/users/__mocks__/dropbox.ts [hunk 19]
```diff
@@ -0,0 +1,66 @@
+import { team } from 'dropbox';
+
+export const membersListFirstPage = (users: number[]) =>
```
Comment (elbalexandre) [inline apps/dropbox/src/app/api/inngest/functions/users/__mocks__/dropbox.ts:3]:
```suggestion
export const membersListFirstPage = (userIds: number[]) =>
```

File: apps/dropbox/drizzle/0000_quiet_sauron.sql (hunk 4 above)
Comment (athielba) [reply apps/dropbox/drizzle/0000_quiet_sauron.sql:10]: make sense

File: apps/dropbox/next.config.js (hunk 6 above)
Comment (athielba) [reply apps/dropbox/next.config.js:7]: While I was deploying to production, I was having this error, this  code fixed the error

File: apps/dropbox/src/database/migrate.ts (hunk 10 above)
Comment (athielba) [reply apps/dropbox/src/database/migrate.ts:18]: I will keep it for now, we can decide it later

File: apps/dropbox/src/common/__mocks__/elba.ts [hunk 20]
```diff
@@ -0,0 +1,25 @@
+import { vi } from 'vitest';
+
+const mocks = vi.hoisted(() => {
+  return {
+    updateUsers: vi.fn(),
+    deleteUsers: vi.fn(),
+  };
+});
+vi.resetModules();
+vi.mock('@elba-security/sdk', () => {
```
[the remark this answers is not in the document]
Comment (athielba) [reply apps/dropbox/src/common/__mocks__/elba.ts:10]: I was having some issue while mocking, thats why I created my own, it works pretty well with my test

File: apps/dropbox/src/app/oauth/service.ts (hunk 16 above)
Comment (athielba) [reply apps/dropbox/src/app/oauth/service.ts:72]: Removed

File: apps/dropbox/src/app/api/inngest/functions/users/run-user-sync-jobs.ts (hunk 17 above)
Comment (athielba) [reply apps/dropbox/src/app/api/inngest/functions/users/run-user-sync-jobs.ts:39]: if you check the line 45, `users.nextCursor` ,is used

File: apps/dropbox/src/app/api/inngest/functions/tokens/run-refresh-tokens.ts [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+  event,
+  step,
+}: InputArgWithTrigger<'tokens/run-refresh-token'>) => {
+  const { organisationId } = event.data;
+
+  const response = await step.run('fetch-refresh-token', async () => {
+    const [token] = await getOrganisationRefreshToken(organisationId);
+
+    if (!token) {
+      throw new Error(
```
[the remark this answers is not in the document]
Comment (athielba) [reply apps/dropbox/src/app/api/inngest/functions/tokens/run-refresh-tokens.ts:17]: Nice catch

File: apps/dropbox/package.json (hunk 7 above)
Comment (athielba) [reply apps/dropbox/package.json:19]: I will update before I merge with staging

File: .gitignore [hunk 22]
```diff
@@ -34,3 +34,6 @@ yarn-error.log*
 
 # vercel
 .vercel
+
+# Ignore .vscode settings
+.vscode/settings.json
```
Comment (athielba) [inline .gitignore:39]: It will be removed before merged with staging

File: apps/dropbox/src/app/oauth/route.ts [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+};
+
+const redirectOnSuccess = () => {
+  return NextResponse.redirect(
+    `${env.ELBA_REDIRECT_URL}?source_id=${env.ELBA_SOURCE_ID}&success=true`
+  );
+};
+
+export async function GET(request: NextRequest) {
+  const { searchParams } = new URL(request.url);
```
Comment (elbalexandre) [inline apps/dropbox/src/app/oauth/route.ts:33]:
```suggestion
  const { searchParams } = request.nextUrl;
```

File: apps/dropbox/src/app/oauth/service.ts [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+        data: {
+          organisationId,
+        },
+      },
+      {
+        name: 'dropbox/users.sync_page.triggered',
+        data: {
+          organisationId,
+          isFirstSync: true,
+          syncStartedAt: new Date().toISOString(),
```
Comment (elbalexandre) [inline apps/dropbox/src/app/oauth/service.ts:97]:
```suggestion
          syncStartedAt: Date.now(),
```
We did that for the other integration

File: .gitignore (hunk 22 above)
Comment (elbalexandre) [reply .gitignore:39]: up

File: apps/dropbox/src/app/oauth/service.ts [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+    ]);
+
+    return {
+      success: true,
+    };
+  } catch (error) {
+    throw new Error('An error occurred while saving the Dropbox access token', {
+      cause: error,
+    });
+  }
```
Comment (elbalexandre) [inline apps/dropbox/src/app/oauth/service.ts:109]: The error could happen will sending an event as well. IMO this does not bring lot of values and make the code harder to read.

File: apps/dropbox/src/app/oauth/service.ts [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+          organisationId,
+          isFirstSync: true,
+          syncStartedAt: Date.now(),
+        },
+      },
+    ]);
+
+    return {
+      success: true,
+    };
```
Comment (elbalexandre) [inline apps/dropbox/src/app/oauth/service.ts:104]: Is it useful ?

File: apps/dropbox/src/test-utils/mock-app-route.ts [hunk 27]
```diff
@@ -0,0 +1,33 @@
...
+  url,
+  handler,
+  body = {},
+  cookies = {},
+}: {
+  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
+  handler: (req: NextRequest) => Promise<void | NextResponse<unknown>>;
+  url: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  body?: any;
```
Comment (elbalexandre) [inline apps/dropbox/src/test-utils/mock-app-route.ts:14]:
```suggestion
  body?: Record<string, string>;
```

Comment (elbalexandre) [inline apps/dropbox/src/test-utils/mock-app-route.ts:14]:
```suggestion
  body?: unknown;
```
or 
```suggestion
  body?: Record<string, unknown>;
```

File: apps/dropbox/src/test-utils/token.ts [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+
+export const insertTestAccessToken = async (tokenDetails: Token[] = defaultAccessToken) => {
+  try {
+    return db.insert(organisations).values(tokenDetails).returning({
+      organisationId: organisations.organisationId,
+      updatedAt: organisations.updatedAt,
+    });
+  } catch (error) {
+    throw error;
+  }
```
Comment (elbalexandre) [inline apps/dropbox/src/test-utils/token.ts:24]:
```suggestion
```

File: apps/dropbox/src/inngest/functions/tokens/data.ts [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+        updatedAt: organisations.updatedAt,
+      });
+  } catch (error) {
+    throw Error(
+      `Not able to update the token details for the organisation with ID: ${organisationId}`,
+      {
+        cause: error,
+      }
+    );
+  }
```
Comment (elbalexandre) [inline apps/dropbox/src/inngest/functions/tokens/data.ts:48]:
```suggestion
```
Are drizzle error not explicit enough ?

File: apps/dropbox/src/connectors/dropbox/dbx-auth.ts [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+      const response = await this.dbxAuth.getAccessTokenFromCode(this.redirectUri, code);
+      const dropboxAuthResult = response.result as DropboxAuthResult;
+
+      return {
+        status: response.status,
+        result: dropboxAuthResult,
+      };
+    } catch (error) {
+      throw error;
+    }
```
Comment (elbalexandre) [inline apps/dropbox/src/connectors/dropbox/dbx-auth.ts:37]:
```suggestion
```

File: apps/dropbox/src/app/page.tsx [hunk 31]
```diff
@@ -0,0 +1,3 @@
+export default function Page() {
+  return <a href="https://admin.elba.ninja">connect elba to Dropbox</a>;
```
Comment (elbalexandre) [inline apps/dropbox/src/app/page.tsx:2]:
```suggestion
  return <main />;
```

File: apps/dropbox/src/inngest/functions/tokens/data.ts (hunk 29 above)
Comment (athielba) [reply apps/dropbox/src/inngest/functions/tokens/data.ts:48]: OK, I am removing the custom error now, will be added lated if I am not happy with the default error details

File: packages/sdk/src/elba.test.ts [hunk 32]
```diff
@@ -25,7 +25,7 @@ describe('users', () => {
       }));
       const elba = new Elba(options);
       await expect(elba.users.update({ users })).resolves.toStrictEqual({
-        success: true,
+        status: 'completed',
```
Comment (elbalexandre) [inline packages/sdk/src/elba.test.ts:28]: ??

Comment (athielba) [reply packages/sdk/src/elba.test.ts:28]: oh no, Find and replaced all, sorry I didn't noticed 🙈

Comment (athielba) [reply packages/sdk/src/elba.test.ts:28]: Good catch

File: apps/dropbox/src/app/install/route.ts [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+  if (!organisationId || !region) {
+    redirect(`${env.ELBA_REDIRECT_URL}?error=internal_error`);
+  }
+
+  cookies().set('state', organisationId);
+  cookies().set('organisation_id', organisationId);
+  cookies().set('region', region);
+
+  const authUrl = await getDropboxAuthUrl(organisationId);
+
```
Comment (elbalexandre) [inline apps/dropbox/src/app/install/route.ts:33]: This function need to handle the unauthorized case (error) + a global try/catch.
Otherwise the user will not be redirected to elba.

File: apps/dropbox/src/app/oauth/route.ts [hunk 34]
```diff
@@ -0,0 +1,85 @@
...
+        errorCode: 'internal_error',
+        organisationId,
+      });
+    }
+
+    await generateAccessToken({ code, organisationId, region });
+
+    return redirectOnSuccess();
+  } catch (error) {
+    return redirectOnError({
```
Comment (elbalexandre) [inline apps/dropbox/src/app/oauth/route.ts:79]: If the error come from dropbox + indicate that we are unauthorized to get a token we should redirect the user with error = unauthorized

File: apps/dropbox/src/inngest/functions/tokens/refresh-token.ts [hunk 35]
```diff
@@ -0,0 +1,78 @@
...
+
+    const tokenDetails = {
+      organisationId,
+      expiresAt,
+      accessToken: await encrypt(accessToken),
+    };
+
+    await updateDropboxTokens(tokenDetails);
+
+    return tokenDetails;
```
Comment (elbalexandre) [inline apps/dropbox/src/inngest/functions/tokens/refresh-token.ts:41]: IMO we should just return `expiresAt`, other fields are not used here.

File: apps/dropbox/src/inngest/functions/users/sync-user-page.test.ts [hunk 36]
```diff
@@ -0,0 +1,198 @@
...
+    const [result] = setup({
+      organisationId,
+      isFirstSync: true,
+      syncStartedAt,
+    });
+
+    await expect(result).rejects.toBeInstanceOf(DropboxResponseError);
+  });
+
+  test('should call elba delete even if there are no users', async () => {
```
Comment (elbalexandre) [inline apps/dropbox/src/inngest/functions/users/sync-user-page.test.ts:74]: all `if` in test case should be replaced by `when`

File: apps/dropbox/src/app/oauth/route.ts (hunk 34 above)
Comment (athielba) [reply apps/dropbox/src/app/oauth/route.ts:79]: this  one handles except all the above error, dropbox redirect with `access_denied` error when it is denied, it is handled in `try`

File: apps/dropbox/src/app/oauth/route.ts [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+      return NextResponse.redirect(redirectOnError('unauthorized'));
+    }
+
+    if (typeof code !== 'string' || state !== cookieState || !organisationId || !region) {
+      return NextResponse.redirect(redirectOnError('internal_error'));
+    }
+
+    await generateAccessToken({ code, organisationId, region });
+  } catch (error) {
+    logger.warn('Could not setup organisation after Dropbox redirection', {
```
Comment (elbalexandre) [inline apps/dropbox/src/app/oauth/route.ts:26]: If `generateAccessToken` somehow throw a 401 we are not redirecting with unauthorized.
I got this feel dumb/redondant in dropbox context, this integration will serve as example.

Comment (athielba) [reply apps/dropbox/src/app/oauth/route.ts:26]: The callback url will have unauthorized param if the user or the account is unauthorized, I don't think we will have another unauthorized once we received the sugccessful callback url, we could handle this once in the catch block, but IMO, it is not necessary

Comment (athielba) [reply apps/dropbox/src/app/oauth/route.ts:27]: @elbalexandre  It is handled [here](https://github.com/elba-security/elba-security/pull/42/files#diff-2d1df1a5e177781f5dc6bddfa0df85dfb69ec4cb07c47fd3cc9e5c27306fd65aR31)

**Review by elbalexandre — APPROVED** [review]
